testing/decorators.py

```diff
@@ -0,0 +1,223 @@
+########################################################################################
+# This file is part of exhale.  Copyright (c) 2017-2018, Stephen McDowell.             #
+# Full BSD 3-Clause license available here:                                            #
+#                                                                                      #
+#                https://github.com/svenevs/exhale/blob/master/LICENSE                 #
+########################################################################################
+
+"""
+The decorators module defines useful class / function decorators for test cases.
+"""
+
+import os
+from copy import deepcopy
+from inspect import isclass
+
+import pytest
+
+from .utils import deep_update
+
+
+__all__ = ['default_confoverrides', 'confoverrides', 'no_run']
+
+
+def _apply_confoverride_to_class(cls, config, priority):
+    """
+    Apply a configuration override ``config`` to class ``cls`` with a given priority.
+
+    We need the priority trick as the default configuration is applied before a possible
+    class-wide decorator, which should supersede the default configuration.  We use
+    ``pytest.mark.exhale`` as a store of ``kwargs`` to apply to ``pytest.mark.sphinx``,
+    and we use the priority to combine these kwargs with respect to priorities.
+
+    This method performs the ultimate ``pytest.mark.sphinx``, saving the ``kwargs`` for
+    this in the ``cls``'s :data:`testing.base.ExhaleTestCase.func_to_sphinx_map`.
+
+    **Parameters**
+        ``cls`` (Subclass of :class:`testing.base.ExhaleTestCase`)
+            The class to apply the ``confoverride`` to.
+
+        ``config`` (:class:`python:dict`)
+            The dictionary of ``confoverrides`` as would be supplied to
+            ``pytest.mark.sphinx``.  These are the overrides to ``conf.py``.
+
+        ``priority`` (:class:`python:int`)
+            The positive integer indicating the priority of the new ``config`` updates,
+            higher values take precedence over lower values.
+
+            For example, when :func:`testing.decorators.default_confoverrides` calls
+            this function, the priority is ``1``.  When the decorator
+            :func:`testing.decorators.confoverrides` is applied to a class, the priority
+            is ``2``.  Finally, when :func:`testing.decorators.confoverrides` is applied
+            to a test function, its priority is ``3``.  Thus the function-level override
+            has the highest priority, meaning any conflicting values with lower level
+            priorities will lose out to the function-level override.
+
+    **Return**
+        Subclass of :class:`testing.base.ExhaleTestCase`
+            The input ``cls`` is returned.
+    """
+    # If not explicitly overriden here, somehow the `func_to_sphinx_map` is getting
+    # populated with previous cls entries.
+    cls.func_to_sphinx_map = {}
```

Comment:
@tkhyn what do you think of the new strategy (see `testing/conftest.py`'s setup and teardown methods).  The only thing I didn't understand is why I had to reset to `{}` for every time this method was called, but as far as I understand this is exactly what should be done.

The reason it was confusing was because the test methods defined in a separate class `CMathsTestsNoRun` were somehow getting in here if I don't reset...

Reply:
I actually had started working on keeping the ``testroot`` stuff in ``ExhaleTestMetaclass.__new__`` so that everything could stay at the same place and there would be no need for pytest setup and teardown hooks.

I finally made it work tonight after a little bit of effort and I think it's significantly cleaner. Check out https://github.com/tkhyn/exhale/blob/testing-fixture-only/testing/base.py#L135

Reply:
WOW that's exactly what I tried to do but totally failed at!  That commit has been added here with some other small changes / docs updates!

Reply:
TBH I'm very pleased with this.  If you want to clean it up go for it, but I just want to build out one small CPP test case and get the class hierarchy tested (or at least a very basic version) and then put this down for a while.  I need to fix the windows stuff and then table exhale for a while :S

Reply:
yeah nah it can stay like that, and once everything will be settle we can worry about making things super-clean :grinning: